cli/ops/mod.rs

```diff
@@ -22,302 +24,113 @@ mod resources;
-    }
-    OP_KILL => {
-      dispatch_json::dispatch(process::op_kill, state, control, zero_copy)
-    }
-    OP_CHDIR => {
-      dispatch_json::dispatch(fs::op_chdir, state, control, zero_copy)
-    }
-    OP_MKDIR => {
-      dispatch_json::dispatch(fs::op_mkdir, state, control, zero_copy)
-    }
-    OP_CHMOD => {
-      dispatch_json::dispatch(fs::op_chmod, state, control, zero_copy)
-    }
-    OP_CHOWN => {
-      dispatch_json::dispatch(fs::op_chown, state, control, zero_copy)
-    }
-    OP_REMOVE => {
-      dispatch_json::dispatch(fs::op_remove, state, control, zero_copy)
-    }
-    OP_COPY_FILE => {
-      dispatch_json::dispatch(fs::op_copy_file, state, control, zero_copy)
-    }
-    OP_STAT => dispatch_json::dispatch(fs::op_stat, state, control, zero_copy),
-    OP_READ_DIR => {
-      dispatch_json::dispatch(fs::op_read_dir, state, control, zero_copy)
-    }
-    OP_RENAME => {
-      dispatch_json::dispatch(fs::op_rename, state, control, zero_copy)
-    }
-    OP_LINK => dispatch_json::dispatch(fs::op_link, state, control, zero_copy),
-    OP_SYMLINK => {
-      dispatch_json::dispatch(fs::op_symlink, state, control, zero_copy)
-    }
-    OP_READ_LINK => {
-      dispatch_json::dispatch(fs::op_read_link, state, control, zero_copy)
-    }
-    OP_TRUNCATE => {
-      dispatch_json::dispatch(fs::op_truncate, state, control, zero_copy)
-    }
-    OP_MAKE_TEMP_DIR => {
-      dispatch_json::dispatch(fs::op_make_temp_dir, state, control, zero_copy)
-    }
-    OP_CWD => dispatch_json::dispatch(fs::op_cwd, state, control, zero_copy),
-    OP_FETCH_ASSET => dispatch_json::dispatch(
-      compiler::op_fetch_asset,
-      state,
-      control,
-      zero_copy,
-    ),
-    _ => panic!("bad op_id"),
-  };
+  // Compiler
+  registry.register_op(OP_NAMESPACE, state.wrap_op(compiler::OpCache));
+  registry
+    .register_op(OP_NAMESPACE, state.wrap_op(compiler::OpFetchSourceFile));
+  registry.register_op(OP_NAMESPACE, state.wrap_op(compiler::OpFetchAsset));
 
-  state.metrics_op_dispatched(bytes_sent_control, bytes_sent_zero_copy);
+  // Errors
+  registry.register_op(OP_NAMESPACE, state.wrap_op(errors::OpFormatError));
+  registry.register_op(OP_NAMESPACE, state.wrap_op(errors::OpApplySourceMap));
 
-  match op {
-    Op::Sync(buf) => {
-      state.metrics_op_completed(buf.len());
-      Op::Sync(buf)
-    }
-    Op::Async(fut) => {
-      use crate::futures::Future;
-      let state = state.clone();
-      let result_fut = Box::new(fut.map(move |buf: Buf| {
-        state.clone().metrics_op_completed(buf.len());
-        buf
-      }));
-      Op::Async(result_fut)
-    }
-  }
+  // Fetch
+  registry.register_op(OP_NAMESPACE, state.wrap_op(fetch::OpFetch));
+
+  // Files
+  registry.register_op(OP_NAMESPACE, state.wrap_op(files::OpOpen));
+  registry.register_op(OP_NAMESPACE, state.wrap_op(files::OpClose));
+  registry.register_op(OP_NAMESPACE, state.wrap_op(files::OpSeek));
```

Comment:
perhaps can pass an array of ops to a function...

Reply:
This doesn't work. The rust compiler doesn't have a way to deduce a concrete type for each element in the array. This normally wouldn't be a problem, but it is when the trait used contains a const field. Here is a example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7bc6caa030b23a0bcdf43188d052c0cb